Add zstd to native access

rjernst opened the pull request:
This commit makes zstd compression available to Elasticsearch. The library is pulled in through maven in jar files for each platform, then bundled in a new platform directory under lib. Access to the zstd compression/decompression is through NativeAccess.

rjernst committed:
Add zstd to native access

This commit makes zstd compression available to Elasticsearch. The
library is pulled in through maven in jar files for each platform, then
bundled in a new platform directory under lib. Access to the zstd
compression/decompression is through NativeAccess.

rjernst committed:
fix distribution build

rjernst committed:
remove debug and leftover

jpountz commented on libs/native/src/main/java/org/elasticsearch/nativeaccess/NoopNativeAccess.java:
```diff
@@ -8,15 +8,24 @@
+    @Override
+    public Zstd getZstd() {
+        logger.warn("cannot compress with zstd because native access is not available");
+        return null;
```
When ZSTD support becomes a requirement in the future, I imagine we'd throw here instead?

rjernst replied:
Yes, ideally there would be no no-op implementation, we would throw at load time.

mark-vieira replied:
Is there a reason we aren't just doing that now?

rjernst replied:
It's contentious, so I want to save that for an isolated discussion/PR.

mark-vieira replied:
This sounds like leniency, which I presumed we had all agreed is abhorrent 😉

jpountz commented on libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkZstdLibrary.java:
```diff
@@ -0,0 +1,99 @@
+    @Override
+    public long compress(ByteBuffer dst, int dstLen, ByteBuffer src, int srcLen, int compressionLevel) {
+        try (Arena arena = Arena.ofConfined()) {
+            var nativeDst = arena.allocate(dstLen);
```
Sometimes, dst may be over-allocated. I wonder if we should allocate `nativeDst` with size min(dstLen, compressBound(srcLen))`. Hopefully the overhead of the additional native call would be negligible.

rjernst replied:
Perhaps we should do that in Zstd.compress so it applies to both impls?

jpountz replied:
Hmm, maybe I'm misunderstanding your suggestion, but it felt to me like this should happen next to where we're allocating the native memory, so it should be in the JDK/JNA impls?

jpountz commented on libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkZstdLibrary.java:
```diff
@@ -0,0 +1,99 @@
+    @Override
+    public long decompress(ByteBuffer dst, int dstLen, ByteBuffer src, int srcLen) {
+        try (Arena arena = Arena.ofConfined()) {
+            var nativeDst = arena.allocate(dstLen);
```
Likewise here for decompression, maybe we should take advantage of the `ZSTD_getFrameContentSize` API to avoid allocating more native memory than necessary.

jpountz replied:
The `compress` method that is exposed right now is guaranteed to only create a single frame: https://github.com/facebook/zstd/blob/dev/lib/zstd.h#L150 so that should always be the case, but you have a point that we may want to decompress buffers compressed by 3rd parties in the future.

We could check if the source length is equal to `ZSTD_findFrameCompressedSize` to check if there is a single frame. But at this point this is becoming complicated enough that it's probably worth deferring to a follow-up PR.

jpountz reviewed (commented):
I can't comment on the build/distribution-related bits that I'm not too familiar with but this looks fantastic!

jpountz commented on libs/native/jna/src/main/java/org/elasticsearch/nativeaccess/jna/JnaZstdLibrary.java:
```diff
@@ -0,0 +1,89 @@
+        assert srcLen != 0;
+        try (Memory nativeDst = new Memory(dstLen);
+             Memory nativeSrc = new Memory(srcLen)) {
+            nativeSrc.write(0, src.array(), src.position(), srcLen);
```
You already have asserts that the buffer is not direct, but since you're using `array()` here, I guess that the buffer needs to not be read-only either. I wonder how much work that would be to make it work transparently across direct, heap and read-only byte buffers. (could be a follow-up)

rjernst replied:
I think I can make it work, it just needs the correct ByteBuffer calls to be backing agnostic.

rjernst committed:
spotless

rjernst committed:
use correct arch

rjernst committed:
Merge branch 'main' into native/zstd

rjernst committed:
better normalize platform

rjernst committed:
Standardize build distribution internals on os/architecture

The build handles platform specific code which may be for arm or x86.
Yet there are multiple ways to describe 64bit x86, and the build
converts between the two in several places. This commit consolidates on
the x64 nomenclature in most places, except where necessary (eg ML still
uses x86_64).

relates #105715

rjernst committed:
adjust x64 rewrite of libs dirs

rjernst committed:
fix platform dir

rjernst committed:
fix packages

rjernst committed:
better os/arch selection and fallback

rjernst committed:
x64 for tests

rjernst committed:
fixup extracted libs dir

rjernst committed:
Merge branch 'main' into native/zstd

rjernst committed:
windows too

rjernst committed:
add tests for jvm options

rjernst committed:
some build tests

rjernst committed:
adjust build tests more

rjernst committed:
force empty commit

rjernst committed:
fix detached configuration with configuration cache

rjernst committed:
add closeable bytebuffer and rework zstd tests

rjernst commented:
@jpountz I added an api to get a buffer that is closeable in 57f5d28281fec1184b4cb5a59421c372d97d55e4.

rjernst committed:
spotless

rjernst committed:
fix tests

rjernst committed:
spotless

rjernst committed:
another try

rjernst committed:
get separator from nio

rjernst committed:
Merge branch 'main' into native/zstd

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

rjernst commented:
Tests are now passing, this is ready for review.

rjernst committed:
make tests work from intellij

rjernst committed:
don't support 0 length buffers

rjernst committed:
spotless

breskeby reviewed (changes requested):
this is pretty cool. some minor remarks/questions

ldematte commented on libs/native/jna/src/main/java/org/elasticsearch/nativeaccess/jna/JnaNativeLibraryProvider.java:
```diff
@@ -8,13 +8,26 @@
-        super("jna", Map.of(PosixCLibrary.class, JnaPosixCLibrary::new));
+        super(
+            "jna",
+            Map.of(
```
Nit: if we think we will continue to add libraries, maybe Map.ofEntries could produce a better formatting here.

rjernst replied:
I'll consider changing that in a followup, you are right it might look nicer, and there are several more of these to come.

ldematte commented on libs/native/src/main/java/org/elasticsearch/nativeaccess/CloseableByteBuffer.java:
```diff
@@ -0,0 +1,18 @@
+public interface CloseableByteBuffer extends AutoCloseable {
+    ByteBuffer buffer();
+
+    @Override
```
Just a curiosity on my side: why override close here?

rjernst replied:
The close method on AutoCloseable throws Exception. By overriding it here I reduce the checked exceptions (to none). Otherwise try-with-resources blocks that use CloseableByteBuffer would have to catch Exception, even though the two implementations don't actually throw anything.

ldematte replied:
That's thoughtful, nice!

ldematte commented on libs/native/src/main/java/org/elasticsearch/nativeaccess/AbstractNativeAccess.java:
```diff
@@ -10,18 +10,36 @@
-    protected AbstractNativeAccess(String name) {
+    protected AbstractNativeAccess(String name, NativeLibraryProvider libraryProvider) {
         this.name = name;
+        this.javaLib = libraryProvider.getLibrary(JavaLibrary.class);
```
I'm a bit confused by where things are placed in the hierarchy; 
I think the rationale is that we implement as much as possible of the `NativeAccess` interface here, delegating to the library provider which will use the correct Java native "bridge" (JNA or JDK). 
And we leave out (implement in derived classes) only those functions that differ across OS/library? (e.g. are implemented by different `NativeLibrary`es?).
Is that correct or am I missing something?

rjernst replied:
Your summary is correct.

This "JavaLibrary" is a special case that I didn't have a better terminology for. It's not actually a library, but it's something that we need to differ between the two bridge implementations, namely how do you get a native buffer that can be closed.

mark-vieira replied:
Yeah, that naming is a bit weird. Would something like `ByteBufferLibrary` make more sense?

rjernst replied:
I named it using "Java" because there will likely be other things we want to differentiate, it's not just for ByteBuffer. I wanted a future place for any java related code with different impls for jna vs jdk.

mark-vieira replied:
I wonder if `StandardLibrary` might be a good alternative. Consider this the end of my bike shedding, I'm also good with this as-is 😉

rjernst replied:
I'm going to keep it JavaLibrary for now, but will consider renaming in the future if/as we add more to it.

ldematte reviewed (commented):
LGTM, but I just looked at the Java part - cannot say about the gradle/groovy side, but I see Rene is already on it :)

rjernst commented on build-tools-internal/src/integTest/groovy/org/elasticsearch/gradle/internal/JdkDownloadPluginFuncTest.groovy:
```diff
@@ -99,17 +99,13 @@ class JdkDownloadPluginFuncTest extends AbstractGradleFuncTest {
-            subprojects {
+        """
+        3.times {
+            subProject(':sub-' + it) << """
```
I had issues with these tests failing due to tasks not being found. It appeared to have to do with using the subproject notation and filtering to the "sub-" prefix (otherwise it picks up the native libs project). So I made the test setup the subproject build files directly instead of configuring through the parent. I opened https://github.com/elastic/elasticsearch/pull/105998 to isolate these changes.

rjernst commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java:
```diff
@@ -59,6 +65,7 @@ public void apply(Project project) {
         configureConfigurations(project);
         configureCompile(project);
         configureInputNormalization(project);
+        configureNativeLibraryPath(project);
```
Unfortunately no unit tests will work without this. BootstrapForTesting initializes NativeAccess, so everything native we use must be available for all tests.

mark-vieira replied:
No _unit_ tests, but we actually disable the security manager for REST tests. Not sure it's worth being conditional though as it doesn't seem problematic to pass this to those tests even if unnecessary.

ChrisHegarty replied:
++ for setting unconditionally. I'm already working on other vector-level related optimisation that will need similar.

rjernst commented on libs/native/libraries/build.gradle:
```diff
@@ -0,0 +1,47 @@
+}
+
+artifacts {
+  runtimePath extractLibs
```
The intent here is to be the one place additional native libs are added. By adding new dependencies above, they will get extracted here. So it is intentionally generic, its extracting all native libs we have (or will have).

breskeby replied:
ok. thanks for clarification

rjernst committed:
address feedback

ldematte reviewed (approved):
Don't know if you want to get additional feedback, but as far as I am concerned looks good!

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java:
```diff
@@ -165,6 +172,25 @@ public static void configureInputNormalization(Project project) {
     }
 
+    private static void configureNativeLibraryPath(Project project) {
+        String nativeProject = ":libs:elasticsearch-native:libraries";
```
Just a heads up that we will need to make some changes to the serverless build to replace this dependency.

mark-vieira commented on libs/native/libraries/build.gradle:
```diff
@@ -0,0 +1,47 @@
+}
+
+def extractLibs = tasks.register('extractLibs', Copy)
+extractLibs.configure {
```
nit: we don't need to separate out the `configure` bit here. This can be simplified to `def extractLibs = tasks.register('extractLibs, Copy) { // config }`.

rjernst replied:
I need the TaskProvider to use in artifacts

mark-vieira replied:
The method accepting a config closure also returns a `TaskProvider`.

mark-vieira commented on libs/native/libraries/build.gradle:
```diff
@@ -0,0 +1,47 @@
+extractLibs.configure {
+  from configurations.libs
+  into layout.buildDirectory.dir('platform')
+  // TODO: fix architecture in uploaded libs
```
What is referring to these? Is it just `SystemJvmOptions.findLibraryPath()`? If so, couldn't we just adapt that?

rjernst replied:
Since we are producing the jar, we should structure the platforms with the same names we already use in the build.

mark-vieira replied:
The jars are packaged such that they align with the conventions used by `Native.load()`. We can change them, but it seems to me it would make more sense for the build to align with existing Java conventions for platform naming.

rjernst replied:
I don't know what Native.load is doing, but that is jna specific, it's not any java conventions. System.loadLibrary only looks directly on java.library.path directly, it doesn't do any recursion, and it's not looking for any platform named dirs. Since we put a platform specific dir on the library path (which is both jna and java library path), I think we should make constructing that path as easy as possible.

mark-vieira replied:
Got it. We can restructure this at the same time we bundle all the native binaries into a single artifact.

mark-vieira commented on libs/native/libraries/build.gradle:
```diff
@@ -0,0 +1,47 @@
+  runtimePath
+}
+
+var zstdVersion = "1.5.5"
```
Can we put this in `version.properties`?

rjernst replied:
I don't think we should. version.properties IMO should be where dependencies that need to be aligned across projects should go. Yet zstd is local to this project, we don't need it anywhere else. IMO we should be removing things from version.properties because it gives the false sense that we should be aligning everything when it's both not necessary in most cases, and tends to force transitive dependency versions that were not meant to be used.

rjernst committed:
Merge branch 'main' into native/zstd

rjernst committed:
use named config

rjernst committed:
use defualt artifact and move maven config to libraries project

mark-vieira commented:
@rjernst I'm wondering if we need some packaging tests here. As it is, there's nothing that's actually exercising zstd in a packaged distribution. Presumably, we'll add functionality that will be executed via REST tests and maybe that'll be good enough?

rjernst commented:
> Presumably, we'll eventually add functionality that relies on zstd that will be executed via REST tests and maybe that'll be good enough?

Yes, that's what I've been expecting.

mark-vieira committed:
Fix distribution artifact resolution

rjernst committed:
Merge branch 'main' into native/zstd

rjernst committed:
spotless

rjernst committed:
cleanup

rjernst committed:
fix build tests

rjernst committed:
give base a name

rjernst committed:
Merge branch 'main' into native/zstd

rjernst commented on libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkCloseableByteBuffer.java:
```diff
@@ -0,0 +1,34 @@
+    private final ByteBuffer bufferView;
+
+    JdkCloseableByteBuffer(int len) {
+        this.arena = Arena.ofShared();
```
A confined arena restricts usage to the current thread, so the buffer could not be used in a different thread.

rjernst commented on libs/native/src/main/java/org/elasticsearch/nativeaccess/Zstd.java:
```diff
@@ -0,0 +1,73 @@
+    public int compress(ByteBuffer dst, ByteBuffer src, int level) {
+        Objects.requireNonNull(dst, "Null destination buffer");
+        Objects.requireNonNull(src, "Null source buffer");
+        long ret = zstdLib.compress(dst, src, level);
```
Yes, I have some of that checked in the impls, but here would be better.

rjernst committed:
move/add buffer checks

rjernst committed:
better packaging test debugging

rjernst committed:
fix library path

rjernst committed:
spotless

rjernst committed:
use nio

rjernst committed:
more debugging

rjernst committed:
use warn so it's easier to find

rjernst committed:
path separator

rjernst committed:
spotless

rjernst committed:
Merge branch 'main' into native/zstd

rjernst committed:
remove debug
